Feature: Adds Lightbox support for the File field

zhephyn opened the pull request:
# Description
Adds Light box support for the file field

Fixes #1647 

# Checkbox
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the [documentation](https://github.com/avo-hq/avodocs)
- [ ] I have added tests that prove my fix is effective or that my feature works

zhephyn commented:
@Paul-Bob I've successfully rebased the PR to remove the unnecessary previous commits. Now the PR, shows one commit which is related to me implementing a custom light box controller that we'll use to implement the feature as opposed to the one i had imported from the stimulus components library before. 

Still though, I'm having trouble registering the controller even after making it available for the cover photo component. 

As for the failing tests, I wanted to first make sure that the functionality is working, then I'll solve them and also add tests for the light box functionality as well.

PS: Sorry about the cluttering of the PR with code climate comments. Not really sure how to delete them to make the PR cleaner.

zhephyn commented:
I expect this message to be logged to the browser console to signify that the controller connected successfully. 
```
connect() {
    console.log("Lightbox controller connected")
  }
```
Usually that's how I do it. Is there a better way for me to check that the controller is regsitered?

Paul-Bob commented:
That is all right, and what page are you visiting and expecting the message to be present on the browser console? I moved to this branch locally and I was able to see the connect debug message on my browser console

zhephyn commented:
I've been navigating to the posts show page for an individual post as every post has a cover photo. So I figured this page would be perfect for seeing the console message.

Paul-Bob commented:
I think you are confusing components, the cover where you added the controller is the one rendered here https://github.com/avo-hq/avo/blob/main/app/components/avo/panel_component.html.erb#L2

And it is the big cover photo that appear on show view like this: https://main.avodemo.com/avo/resources/events/47

zhephyn commented:
Not sure if I get what you're trying to say. I had added the controller to the opening div inside the ```cover_photo_component.html.erb```
Was the controller supposed to be added to a different file and is this not the component which renders the cover photo on the post show page?

Paul-Bob commented:
> Not sure if I get what you're trying to say. I had added the controller to the opening div inside the `cover_photo_component.html.erb`

That component is rendered when the resource has `self.cover_photo` configured.

> Was the controller supposed to be added to a different component file and is this not the component which renders the cover photo on the post show page?

I'm unclear on where the controller is supposed to go. Based on the original issue request, I think it should be added to the file and files fields.

What I'm trying to say is that you added the controller to a component that isn't present on the post resource's show component. The `cover_photo_component.html.erb` is only responsible for rendering this:

```ruby
class Avo::Resources::Event < Avo::BaseResource
  self.cover_photo = {
    size: :xl,
    visible_on: [:show, :index],
    source: -> {
      if record.present?
        record.cover_photo
      else
        Event.first&.cover_photo
      end
    }
  }
end
```

https://main.avodemo.com/avo/resources/events/47
![image](https://github.com/user-attachments/assets/0d80cfe2-4630-4720-8f24-5dd8d3588313)

Paul-Bob commented:
```ruby
field :cover_photo, as: :file, ...
field :cover_photo, as: :external_image, ...
```

I think I understand from where your confusion started, this fields have nothing to do with the cover component, these are just 2 fields, one as file and the other as external image, that are using the cover_photo name by coincidence.

Paul-Bob commented:
> Ohhh I get it now. let me dig around and see if I can figure out where it's supposed to go. Gonna report back ASAP!

Cool! 

No rush, take your time

Paul-Bob commented:
> Hey @Paul-Bob
> 
> So I've finally moved the data-action and target attributes for the lightbox stimulus controlller to the right file which renders the cover photo on the post show page. This file is the `grid_item_component.html.erb`.

Why are you adding it to the `grid_item_component.html.erb`? It is indeed the component that renders the cover photo on the post resource but on index when on grid view, not on show page. This implementation doesn't require any modification to that component only to the files components ( let me know if you need me to show which ones I'm talking about ).

Could you let me know why are you registering the controller on this specific component? Probably if I understand the root cause of the confusion I could clarify it.

> At this point, I expected the functionality to work, but just like before, its not working. Still not sure if the controller was registered because even when i added `data-controller="lightbox"` to the main div of the `grid_item_component.html.erb` file, i expect to see a message logged to the console when this component is rendered like you had mentioned. However, that's not the case.
> 
> All in all, I'm not sure if its not working because of incorrect implementation or because the controller was never registered to begin with.

Let's focus on my question from above then I can help you to debug on the correct file field related component.

zhephyn commented:
So, when a cover_photo field is declared for the post resource like below:
```ruby
    field :cover_photo, as: :file, is_image: true, full_width: true, hide_on: [], accept: "image/*", stacked: true
```
The rendering of the cover_photo is handled by the file field. Inside the file_field folder at this path :
```ruby
/home/zhephyn/Desktop/avo/app/components/avo/fields/file_field
```
We have component files. Since the cover photo is supposed to be rendered on the post show page, we focus on the ```show_component.html.erb``` and ```show_component.rb``` files in the file_field folder. 

Inside the ```show_component.html.erb``` file, we can see that we have this code:
```ruby
<%= field_wrapper(**field_wrapper_args) do %>
  <%= render Avo::Fields::Common::Files::ViewType::GridItemComponent.new resource: @resource, field: @field %>
<% end %>
```
This let's us know that on the post show page, the cover_photo is rendered by the component above which is the ```grid_item_component.html.erb```

This was my line of thinking when determining which file renders the cover photo. Not sure if that's the right line of thinking though.

As for why I'm registering the controller in this component, I wanted stimulus to know that the image rendered by the ```grid_item_component.html.erb``` component, in this case cover_photo for the post resource, is what I'm targeting. That's why i added the data-target attribute to it. Such that when this image is clicked on, it will be displayed inside the lightbox partial that I made available globally throughout the application as part of the layout file. That's why i added the stimulus attributes to the image tag of that component. 

For the times I've worked with stimulus, usually that's its done. You figure out what part of the page you want to make interactive, register a controller inside a div wrapping that part of the page, then specify your target attributes.

Paul-Bob commented:
Got it, I am the one confused 🤦🏼‍♂️ 

You're on the right component, I confounded it with the `app/components/avo/index/grid_item_component.html.erb` 😆 

I went to this branch, run the server, visited a post and observed this on the web console:

<img width="1917" height="1078" alt="image" src="https://github.com/user-attachments/assets/3de5c6cd-133e-43fb-9d1c-d9e9196bcf51" />

I'm not sure what is happening on your side, I have a wild guess, how are you running the server, are you using `rails s` or `bin/dev`?

zhephyn commented:
Hey @Paul-Bob 

I've pushed  an implementation of the light box functionality which you can try out in the browser by clicking on the cover photo displayed on the post show page. 

I've kept the styling minimal in case we end up having to delete/relocate some files. After confirming that everything is where its supposed to be, I'll proceed with the styling and the tests. 

Kindly take a look at it whenever its convenient for you and lmk in case of any changes. 

Thanks!

Paul-Bob commented:
@zhephyn why are there +9,406 −6,462 changes?

zhephyn commented:
I'm not sure about that large diff file, I had rebased it and it wasn't showing up in the commit history anymore. Let me clean it up and get back to you!

Paul-Bob reviewed (changes requested):
Let's refine the styling. The images are currently stretched, and it doesn't quite look right.

https://www.loom.com/share/41091431f7664b018961dfbb52095752?sid=8192140a-e087-4e95-b3ac-3323432dcb31

zhephyn commented:
Thanks for the review. Gonna work on it ASAP!

zhephyn commented:
hey @Paul-Bob Any resource you'd recommend for debugging tail wind? 

I've tried to fix the image stretch, however on making changes the functionality ends up breaking. One idea I was counting on, was to use the `object-contain` class, however that didn't work. Now I'm stuck

Paul-Bob commented:
Hi @zhephyn  

I usually use the browser's dev tools to "debug Tailwind".  
What exactly do you want to debug? Are the styles you're writing not being applied in the browser?  
For example, if you add `bg-[#50d71e]` or any other custom color, does it apply after a refresh?

zhephyn commented:
Hey @Paul-Bob . Ideally, I should be able to add some padding around the image such that it's displayed in the center of the page, plus making the edges of the image a bit curved, also maybe make the background of the image a bit blurry. However, when I add styles to implement any of these changes, the layout just breaks. Then maybe I though increasing the z-index is the issue but even that doesn't work. None of the styles are applied

Paul-Bob commented:
> So, the green background is not applied when I add it.

Are you using the `bin/dev` command to start the server?

zhephyn commented:
Hey @Paul-Bob, The github actions bot just marked this as stale, which may result in the PR being closed. I could try again to debug the UI tailwind issue if its cool with you.

Paul-Bob commented:
> I could try again to debug the UI tailwind issue if its cool with you.

The issue seems to be on your local environment. `bin/dev` does run `css: yarn build:css --watch`, so I'm not sure why the watcher is not working on your environment. You're the right person to debug and fix it on your side. Alternatively, if you want to move fast without debugging this problem, you can just run `yarn build:css` after each style change to rebuild and confirm the effect in the browser.

Paul-Bob commented:
> Update. Done debugging the watcher issue.

What was the issue?

zhephyn commented:
I actually don't know. but I deleted the project  locally and re-cloned it again from Github. Following that, the watcher was picking up the new classes when i ran `bin/dev`. Gonna now proceed to work on the styles so that we can get this merged!

Paul-Bob commented:
Closing this PR to keep the queue clean since we are not focusing on it right now.

We will probably reopen and rebase it on Avo 4.

Thank you @zhephyn for all the great work on this one!
